src/plugins/plugin.htmllegend.js

```diff
@@ -0,0 +1,301 @@
+
+				container.appendChild(element);
+			});
+
+			if (output === 'HTML') {
+				container = container.outerHTML; // output html by default for legacy purposes
+			}
+			return container;
+		},
+
+		/**
+		 * Returns a HTML legend (wrapper for generateLegend)
+		 * @param {DOM node/*} defaultTarget - The defaultTarget or optional parameter if not DOM node.
+		 * @param {*} optional - The optional arguments from `generateLegend()`, to be passed to callback function.
+		 * @returns {String/DOM node} HTML legend string/node.
+		 */
+		generate: function(defaultTarget) {
+			var options = this.chart.options.htmllegend;
+			var legendFn;
+
+			if (typeof this.chart.options.legendCallback === 'function') {
+				legendFn = this.options.legendCallback;
+				console.warn('options.legendCallback is deprecated, replace by options.htmllegend or options.htmllegend.callback');
+			} else {
+				legendFn = this.create;
+				if (typeof options === 'function') {
+					legendFn = options;
+					options = undefined;
+				} else if (options && typeof options.callback === 'function') {
+					legendFn = options.callback;
+				}
+			}
+
+			// remove default target from optional arguments
+			var args = defaultTarget && defaultTarget.nodeType === Node.ELEMENT_NODE
+				? Array.prototype.slice.call(arguments, 1)
+				: Array.prototype.slice.call(arguments);
+			var legend = legendFn.call(this, this.chart, options, args);
+
+			// attach legend to DOM and bind event listeners
+			var legendNode = legend;
+			var target = optionOrDefault(options, 'target', defaultTarget);
+			if (target && target.nodeType === Node.ELEMENT_NODE) {
+				if (typeof legend === 'string') {
+					target.innerHTML = legend;
+					legendNode = target.childNodes[0];
+				} else {
+					target.appendChild(legend);
+				}
+
+			}
+
+			// save a link to the legend to fire afterUpdate event
+			if (legendNode.nodeType === Node.ELEMENT_NODE) {
+				var listeners = optionOrDefault(options, 'listeners');
+				this.bindEventListeners(legendNode, listeners);
+
+				this.chart.htmllegend.container = legendNode;
+			}
+
+			return legend;
+		},
+
+		/**
+		 * Binds event listeners to legend
+		 * @param {Object} node - The DOM node
+		 * @param {Object} node - The listeners object
+		 * @returns undefined
+		 */
+		bindEventListeners: function(node, listeners) {
+			if (node.nodeType === Node.ELEMENT_NODE) {
+				helpers.each(listeners, function(listener, type) {
+					var options = optionOrDefault(listener, 'options', false);
+					listener = optionOrDefault(listener, 'listener', listener);
+
+					node.addEventListener(type, listener, options);
+				});
+			}
+		}
+	});
+
+
+	function createNewLegendAndAttach(chart, legendOpts) {
+		var htmllegend = new Chart.HTMLLegend({
+			options: legendOpts,
+			chart: chart,
+			_construct: true // temporary flag to auto-generate legend
+		});
+
+		chart.htmllegend = htmllegend;
+	}
+
+	return {
+		id: 'htmllegend',
+
+		beforeInit: function(chart) {
```

Comment:
When options are scoped under `options.plugins.htmllegend`, you will get the plugin options as the last [hook argument](/chartjs/Chart.js/blob/master/src/core/core.plugin.js#L175)